### PR TITLE
Implement sql deleted database restore

### PR DESCRIPTION
**Pull request by jaredmoo**
Added `az sql db list-deleted` command and `az sql db restore --deleted-time` parameter so that now deleted db's can be found and restored. 

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst with an appropriate description of the change (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

**Comment by jaredmoo** [comment]
```
Command
    az sql db list-deleted: Get a list of deleted database that can be restored.

Arguments
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.

Command
    az sql db restore: Creates a new database by restoring from a database backup.

Arguments
    --dest-name         [Required]: Name of the database that will be created as the restore
                                    destination.
    --name -n           [Required]: Name of the Azure SQL Database.
    --resource-group -g [Required]: Name of resource group. You can configure the default group
                                    using `az configure --defaults group=<name>`.
    --server -s         [Required]: Name of the Azure SQL server.
    --edition                     : The edition for the new database.
    --elastic-pool                : Name of the elastic pool to create the new database in.
    --service-objective           : Name of service objective for the new database.
    --tags                        : Resource tags.

Restore Point Arguments
    --deleted-time                : If specified, restore from a deleted database instead of from an
                                    existing database. Must match the deleted time of a deleted
                                    database in the same server. Either --time or --deleted-time (or
                                    both) must be specified.
    --time -t                     : The point in time of the source database that will be restored
                                    to create the new database. Must be greater than or equal to the
                                    source database's earliestRestoreDate value. Either --time or
                                    --deleted-time (or both) must be specified.

Global Arguments
    --debug                       : Increase logging verbosity to show all debug logs.
    --help -h                     : Show this help message and exit.
    --output -o                   : Output format.  Allowed values: json, jsonc, table, tsv.
                                    Default: json.
    --query                       : JMESPath query string. See http://jmespath.org/ for more
                                    information and examples.
    --verbose                     : Increase logging verbosity. Use --debug for full debug logs.
```

**Comment by jaredmoo** [comment]
@pixia to review from SQL backup/restore team. Thx

**Comment by jaredmoo** [comment]
Looks like sql sdk 0.7.0 is broken, I need to take a fix there in order to fix the security alert test failure.

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py [hunk 1]
```diff
@@ -57,6 +57,13 @@ def __init__(self, database_name, server_name, resource_group_name):
         self.server_name = server_name
         self.resource_group_name = resource_group_name
 
+    def id(self):
+        return '/subscriptions/{}/resourceGroups/{}/providers/Microsoft.Sql/servers/{}/databases/{}'.format(
+            quote(get_subscription_id()),
+            quote(self.resource_group_name),
+            quote(self.server_name),
+            quote(self.database_name))
```
Comment (troydai) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py]: I would rather have this utility put in the SDK.

Comment (troydai) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py]: Not a blocking issue.

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py [hunk 2]
```diff
@@ -178,20 +179,27 @@ def db_create_replica(
...
+        restore_point_in_time=None,
+        source_database_deletion_date=None,
         **kwargs):
 
+    if not (restore_point_in_time or source_database_deletion_date):
+        raise CLIError('Either --time or --deleted-time must be specified.')
+
     # Set create mode properties
-    kwargs['create_mode'] = 'PointInTimeRestore'
+    is_deleted = (source_database_deletion_date is not None)
```
Comment (troydai) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py]: The parenthesis is not needed.

**Comment by troydai** [comment]
@jaredmoo approved with a few comments. Please resolve the merge conflicts and let me know if you'd like this to be merged.

**Comment by troydai** [comment]
Please fix the CI.

**Comment by jaredmoo** [comment]
oops, this exposed another python sdk bug caused by swagger. https://github.com/Azure/azure-rest-api-specs/pull/1383

**Comment by troydai** [comment]
Ok. I assume this will not make sprint 18. Let me know when it is ready to be merged.

**Comment by troydai** [comment]
@jaredmoo is SDK fixed?

**Comment by jaredmoo** [comment]
Server AD Admin test requires a fix in python SDK. This is part of https://github.com/Azure/azure-sdk-for-python/pull/1413

**Comment by troydai** [comment]
@jaredmoo, since the SDK has been updated, can you revive this PR?

**Comment by jaredmoo** [comment]
Reviving

**Comment by troydai** [comment]
image

**Comment by jaredmoo** [comment]
Reviewed by SQL DB backup/restore team PM, need review from @hungmai-msft

**Comment by troydai** [comment]
Changes look good to me. Pending other SQL team member's sign off as well as rebase to master.

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/commands.py [hunk 3]
```diff
@@ -54,7 +55,6 @@
             c.custom_command('restore', 'db_restore', no_wait_param='raw')
             c.command('show', 'get')
             c.custom_command('list', 'db_list')
-            c.command('list-usages', 'list_usages')
```
Comment (hungmai-msft) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/commands.py]: why remove this?

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/test_sql_commands.py [hunk 4]
```diff
@@ -389,48 +413,98 @@ def test_sql_db_restore(self, resource_group, resource_group_location, server):
...
+            deleted_dbs = list(self.cmd('sql db list-deleted -g {} -s {}'.format(rg, server)).get_output_in_json())
+
+            if deleted_dbs:
+                # Deleted db found, stop polling
+                break
+
+            # Deleted db not found, sleep (if running live) and then poll again.
+            if self.is_live:
+                self.assertTrue(datetime.now() < start_time + timeout, 'Deleted db not found before timeout expired.')
+                sleep(10)  # seconds
```
Comment (hungmai-msft) [inline src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/test_sql_commands.py]: we may want a timeout to break out of potential infinite loop

Comment (troydai) [reply src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/tests/test_sql_commands.py]: The timeout is set on line 471 and tested on line 482.

File: src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/commands.py (hunk 3 above)
Comment (jaredmoo) [reply src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/commands.py]: I think github is being dumb about merges. It should look right when I rebase.

Comment (jaredmoo) [reply src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/commands.py]: Never mind, it is really deleted, however sql db list-usages is redefined further down. That later one is the correct one.

**Pull request merged**